File: apps/tracking-api/internal/broker/rabbitmq.go
package broker

import (
	"encoding/json"
	"fmt"
	"log"
	"math"
	"os"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"
)

const (
	exchangeName = "agachadeats_events"
)

type RabbitMQBroker struct {
	conn    *amqp.Connection
	channel *amqp.Channel
}

// Connect starts a connection with exponential backoff
func Connect(url string, maxRetries int) (*RabbitMQBroker, error) {
	if url == "" {
		url = os.Getenv("RABBITMQ_URL")
		if url == "" {
			url = "amqp://guest:guest@rabbitmq:5672/"
		}
	}

	var conn *amqp.Connection
	var err error

	for i := 0; i < maxRetries; i++ {
		log.Printf("Attempting to connect to RabbitMQ (Attempt %d/%d)...", i+1, maxRetries)
		conn, err = amqp.Dial(url)
		if err == nil {
			log.Println("Successfully connected to RabbitMQ")
			break
		}

		log.Printf("Failed to connect: %v", err)
		if i < maxRetries-1 {
			waitTime := time.Duration(math.Pow(2, float64(i))) * time.Second
			log.Printf("Retrying in %v...", waitTime)
			time.Sleep(waitTime)
		}
	}

	if err != nil {
		return nil, fmt.Errorf("failed to connect to RabbitMQ after retries: %w", err)
	}

	ch, err := conn.Channel()
	if err != nil {
		return nil, fmt.Errorf("failed to open a channel: %w", err)
	}

	err = ch.ExchangeDeclare(
		exchangeName, // name
		"topic",      // type
		true,         // durable
		false,        // auto-deleted
		false,        // internal
		false,        // no-wait
		nil,          // arguments
	)
	if err != nil {
		return nil, fmt.Errorf("failed to declare an exchange: %w", err)
	}

	return &RabbitMQBroker{
		conn:    conn,
		channel: ch,
	}, nil
}

func (b *RabbitMQBroker) Close() {
	if b.channel != nil {
		b.channel.Close()
	}
	if b.conn != nil {
		b.conn.Close()
	}
}

// ConsumePingEvents binds an exclusive queue to the exchange and consumes "system.#"
func (b *RabbitMQBroker) ConsumePingEvents() error {
	q, err := b.channel.QueueDeclare(
		"",    // name (empty means a random name will be generated)
		false, // durable
		false, // delete when unused
		true,  // exclusive
		false, // no-wait
		nil,   // arguments
	)
	if err != nil {
		return fmt.Errorf("failed to declare a queue: %w", err)
	}

	// Bind the queue to the exchange
	err = b.channel.QueueBind(
		q.Name,       // queue name
		"system.#",   // routing key
		exchangeName, // exchange
		false,        // no-wait
		nil,          // args
	)
	if err != nil {
		return fmt.Errorf("failed to bind queue: %w", err)
	}

	msgs, err := b.channel.Consume(
		q.Name, // queue
		"",     // consumer tag
		false,  // auto-ack (we use manual ack)
		false,  // exclusive
		false,  // no-local
		false,  // no-wait
		nil,    // args
	)
	if err != nil {
		return fmt.Errorf("failed to register a consumer: %w", err)
	}

	// Idempotency check cache
	seenMsgs := make(map[int64]bool)

	log.Println("Waiting for events. To exit press CTRL+C")

	go func() {
		for d := range msgs {
			var payload struct {
				Msg       string `json:"msg"`
				Timestamp int64  `json:"timestamp"`
			}

			if err := json.Unmarshal(d.Body, &payload); err != nil {
				log.Printf("Error unmarshaling message body: %v", err)
				d.Nack(false, false)
				continue
			}

			if seenMsgs[payload.Timestamp] {
				log.Printf("Duplicate ping event detected: %d. Acknowledging and discarding.", payload.Timestamp)
				d.Ack(false)
				continue
			}

			log.Printf("Received event [RoutingKey: %s]: %s (Timestamp: %d)", d.RoutingKey, payload.Msg, payload.Timestamp)
			seenMsgs[payload.Timestamp] = true
			d.Ack(false)
		}
	}()

	return nil
}

File: apps/orders-api/src/lib/rabbitmq.ts
import * as amqp from "amqplib";

const RABBITMQ_URL =
  process.env.RABBITMQ_URL || "amqp://guest:guest@rabbitmq:5672/";
const EXCHANGE_NAME = "agachadeats_events";

type Connection = Awaited<ReturnType<typeof amqp.connect>>;
type Channel = Awaited<ReturnType<Connection["createChannel"]>>;

let connection: Connection | null = null;
let channel: Channel | null = null;
let connectingPromise: Promise<{ connection: Connection; channel: Channel }> | null = null;

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Connects to RabbitMQ with retry logic.
 * Returns a singleton connection and channel.
 */
export async function connectRabbitMQ(retries = 5): Promise<{ connection: Connection; channel: Channel }> {
  // If we already have a connection and channel, return them
  if (connection && channel) {
    return { connection, channel };
  }

  // If a connection attempt is already in progress, wait for it
  if (connectingPromise) {
    return connectingPromise;
  }

  connectingPromise = (async () => {
    for (let i = 0; i < retries; i++) {
      try {
        console.log(
          `Attempting to connect to RabbitMQ (Attempt ${i + 1}/${retries})...`,
        );
        
        // amqplib promise API
        const conn = await amqp.connect(RABBITMQ_URL);
        const ch = await conn.createChannel();

        // Define "Topic" exchange
        await ch.assertExchange(EXCHANGE_NAME, "topic", { durable: true });

        console.log("Successfully connected to RabbitMQ");

        // Handle connection events
        conn.on("error", (err: Error) => {
          console.error("RabbitMQ connection error:", err);
          connection = null;
          channel = null;
        });

        conn.on("close", () => {
          console.warn("RabbitMQ connection closed.");
          connection = null;
          channel = null;
          connectingPromise = null;
        });

        connection = conn;
        channel = ch;
        return { connection, channel };
      } catch (error) {
        console.error(`RabbitMQ connection failed: ${(error as Error).message}`);
        if (i < retries - 1) {
          const waitTime = Math.pow(2, i) * 1000;
          console.log(`Retrying in ${waitTime / 1000} seconds...`);
          await sleep(waitTime);
        } else {
          connectingPromise = null;
          throw new Error(`Failed to connect to RabbitMQ after ${retries} retries`);
        }
      }
    }
    connectingPromise = null;
    throw new Error("Failed to connect to RabbitMQ");
  })();

  return connectingPromise;
}

/**
 * Publishes an event to the RabbitMQ exchange.
 */
export async function publishEvent(routingKey: string, payload: any) {
  const { channel: ch } = await connectRabbitMQ();
  
  const message = Buffer.from(JSON.stringify(payload));
  const success = ch.publish(EXCHANGE_NAME, routingKey, message, {
    persistent: true,
  });

  if (!success) {
    console.warn("Failed to publish message immediately, buffer might be full");
  }
  return success;
}

/**
 * Closes the RabbitMQ connection and channel.
 */
export async function closeRabbitMQ() {
  if (channel) {
    try {
      await channel.close();
    } catch (err) {
      console.error("Error closing RabbitMQ channel:", err);
    }
    channel = null;
  }
  if (connection) {
    try {
      await connection.close();
    } catch (err) {
      console.error("Error closing RabbitMQ connection:", err);
    }
    connection = null;
  }
  connectingPromise = null;
}
